### PR TITLE
Added typescript definition file

### PR DESCRIPTION
**Pull request by aamirtestable**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    function execute(func: Function)
+
+    namespace dataTable {
+        function open()
+        namespace open{
+            function get(i: number);
+            function random();
+            function next(options?: object);
+        }
```
Comment (avistramer) [inline index.d.ts:37]: Is this declaring the response type of the `open()` function? Should it be a top level type here maybe?

I think we should declare the response types of the methods:

```
function get(i: number): Promise<DataRow>
function random(): Promise<DataRow>
function next(options?: object): Promise<Array<DataRow>>

//...

namespace DataRow {
  index: number;
  data: object;
  indexed: Array<any>;
}
```

Or something like that.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    function results()
+
+    namespace results {
+        function toResourceName(url: string)
+    }
```
Comment (avistramer) [inline index.d.ts:44]: Take a look at `results.js` there are quite a few methods in there. The `toResourceName()` method is then tacked onto the function as a property in addition to many other functions. I followed the first answer [here](https://stackoverflow.com/questions/12766528/build-a-function-object-with-properties-in-typescript) to account for that. I think something like this:

```
var results: { 
  (resource?: string, url?: string): Result; 
  get(options: object): Promise<number>;
  get(name: string, key?: string): Promise<number>;
  waitForCondition(options: Condition): Promise<void>;
  waitForValue(options: ValueCondition): Promise<void>;
  incrementAndWaitForValue(name: string, value: number): Promise<void>;
  barrier(name: string, value: number): Promise<void>;
  toResourceName: (url: string) => string; 
  current: Result;
}

namespace ValueCondition {
  value: number;
  name: string;
  key?: string;
}

namespace Condition {
  condition: (val: number) => boolean;
  name: string;
  key?: string;
}

namespace Result {
  function timing(options: ResultOptions): Promise<void>;
  function timing(name: string, key?: string, val?: number, units?: string): Promise<void>;
  function counter(options: ResultOptions): Promise<void>;
  function counter(name: string, key?: string, val?: number, units?: string): Promise<void>;
  function histogram(options: ResultOptions): Promise<void>;
  function histogram(name: string, key?: string, val?: number, units?: string): Promise<void>;
  function metered(options: ResultOptions): Promise<void>;
  function metered(name: string, key?: string, val?: number, units?: string): Promise<void>;
  function setTraceStatus(status: string): void;
  function markAsSuccess(): void;
  function markAsFailure(): void;
}

namespace ResultOptions {
  name: string;
  key?: string;
  val?: number;
  units?: string;
}
```

Didn't check this at all but something like that.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    function stopwatch(func: Function, param1?: string, param2?: string)
+
+    function waitForFinish()
+
+    var isLocal: boolean;
+    var isSmokeTest: boolean;
+    var info: any;
```
Comment (avistramer) [inline index.d.ts:52]: Might be worth declaring a type instead of `any` to at least document the first level of properties in the `info` object.

**Comment by aamirtestable** [comment]
https://www.typescriptlang.org/docs/handbook/basic-types.html `Array` `number and `string`

**Comment by avistramer** [comment]
https://github.com/basarat/typescript-book/blob/master/docs/styleguide/styleguide.md#class it does look like upper case is the convention. I think a lot of the places where I tried to add `namespace` declarations it should have been a `class` declaration.

**Comment by aamirtestable** [comment]
Well, we do have `testableUtil.log.info` or `testableUtil.dataTable`

So if I use Pascal case, or any other case, typescript will compile but when that compiled code is sent to V8 engine, it will give error that `Log` is undefined. 

Here is a an example log

```
07/16/2020 11:00:42.780 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Downloading files into directory /var/folders/h5/rm5f3t8s777bnw2ffmp3xr3h0000gp/T/tmp-462327vERXwW2NZeG...
07/16/2020 11:00:42.790 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Concurrent user started
07/16/2020 11:00:42.797 PM ERROR [aamir-agents | Aamirs-MacBook-Pro.local] Iteration error
07/16/2020 11:00:42.798 PM ERROR [aamir-agents | Aamirs-MacBook-Pro.local] script.js:24
testableUtils.Log.info('1234');
                  ^

TypeError: Cannot read property 'info' of undefined
    at script.js:24:19

07/16/2020 11:00:42.800 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Concurrent user finished

```

**Comment by avistramer** [comment]
> Well, we do have `testableUtil.log.info` or `testableUtil.dataTable`
> 
> So if I use Pascal case, or any other case, typescript will compile but when that compiled code is sent to V8 engine, it will give error that `Log` is undefined.
> 
> Here is a an example log
> 
> ```
> 07/16/2020 11:00:42.780 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Downloading files into directory /var/folders/h5/rm5f3t8s777bnw2ffmp3xr3h0000gp/T/tmp-462327vERXwW2NZeG...
> 07/16/2020 11:00:42.790 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Concurrent user started
> 07/16/2020 11:00:42.797 PM ERROR [aamir-agents | Aamirs-MacBook-Pro.local] Iteration error
> 07/16/2020 11:00:42.798 PM ERROR [aamir-agents | Aamirs-MacBook-Pro.local] script.js:24
> testableUtils.Log.info('1234');
>                   ^
> 
> TypeError: Cannot read property 'info' of undefined
>     at script.js:24:19
> 
> 07/16/2020 11:00:42.800 PM DEBUG [aamir-agents | Aamirs-MacBook-Pro.local] Concurrent user finished
> ```

You're confusing two completely separate things. Type names from property names. The properties names can be lower case (`testableUtils.log`), the type/interface names should be upper case (`interface Log { ... }`).

**Comment by avistramer** [comment]
I think it's because you're overusing the namespace construct. Inside the module most of the namespaces should really be interfaces.

```
interface Log {

}

var log: Log;
```

Or something like that.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+declare module testableUtils {
+    namespace log {
+        function fatal(message: any);
+        function info(message: any);
+        function error(message: any);
+        function debug(message: any);
+        function trace(message: any);
+    }
+
+    interface events {
```
[the remark this answers is not in the document]
Comment (aamirtestable) [reply index.d.ts:10]: Fixed

File: index.d.ts (hunk 2 above)
Comment (aamirtestable) [reply index.d.ts:44]: Hmm, I need to check this in details. Did you used any generator? TBH this is also bit new for me.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        timing(options: ResultOptions): Promise<void>;
+        timing(name: string, key?: string, val?: number, units?: string): Promise<void>;
+        counter(options: ResultOptions): Promise<void>;
+        counter(name: string, key?: string, val?: number, units?: string): Promise<void>;
+        histogram(options: ResultOptions): Promise<void>;
+        histogram(name: string, key?: string, val?: number, units?: string): Promise<void>;
+        metered(options: ResultOptions): Promise<void>;
+        metered(name: string, key?: string, val?: number, units?: string): Promise<void>;
+
+        data
```
Comment (avistramer) [inline index.d.ts:101]: I think this should be a type with two string fields: resource and url.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
+import {EventEmitter} from "events";
+
+declare module testableUtils {
+    var isLocal: boolean;
+    var isSmokeTest: boolean;
+    var events: EventEmitter;
+
+    function execute(func: Function)
+
+    function stopwatch(func: Function, param1?: string, param2?: string)
```
Comment (avistramer) [inline index.d.ts:10]: Rename to `code`, `metricName`, and `resource` as the parameter names.

Comment (aamirtestable) [reply index.d.ts:10]: My code messed up, i was working inside node module and when I started agent, it reinstalled my typescript folder modules. Had to do all again, and this got missed

**Pull request merged**